accepted/000070-cluster_awareness.md

```diff
@@ -0,0 +1,322 @@
+
+When the import action is selected, Uyuni/SUSE Manager will register the management node to Uyuni/SUSE Manager (by replicating what is done under Systems > Bootstrapping).
+Additionally, Uyuni/SUSE Manager asks for the name and the type of cluster that the user wants to register.
+
+Uyuni/SUSE Manager will store all entered cluster details data into the database.
+
+After successful registration, the cluster overview will be shown.
+
+Example:
+- Uyuni/SUSE Manager knows that it can manage CaaSP, SES, and SHAP clusters (because the cluster provider manager is installed into the same Uyuni/SUSE Manager server)
+- Now the user wants to import the cluster: Uyuni/SUSE Manager asks for the cluster name and the cluster type (from the list gathered above).
+- Uyuni/SUSE Manager will ask the management node and proceed to register it to Uyuni/SUSE Manager (if not already registered to Uyuni/SUSE Manager).
+- Uyuni/SUSE Manager makes sure that the cluster management stack for the imported cluster is installed in the management node.
+
+### Creating a new cluster
+
+When creating a new cluster, Uyuni/SUSE Manager will ask:
+
+1. Cluster name and cluster type
+2. A minion to provision as the management node identified by a specific System Add-On and provisioned with a Salt Formula that will install the cluster management stack on the minion. A list of available minions will be presented.
+3. Uyuni/SUSE Manager will apply the Salt highstate to provision the system as the cluster management stack.
+
+Only the cluster management stack is provisioned on the management node at this time. At the time of writing, nothing forbids that cluster management stack for all types of a cluster can be installed on the same node.
+
+All other optional actions:
+- provisioning a load balancer
+- copying SSH keys from the management node to aspirant nodes of the cluster
+
+will __not__ be managed by Uyuni/SUSE Manager and must be provisioned manually. In a future follow-up, Uyuni/SUSE Manager can make use of [Salt orchestrate runner](https://docs.saltstack.com/en/latest/topics/orchestrate/orchestrate_runner.html#orchestrate-runner) to provision these complex setups before provisioning the management node.
+
+Uyuni/SUSE Manager will store all entered cluster details data into the database and after successful provisioning of the management node, the cluster overview will be shown.
+
+
+### Cluster overview
+
+Under “Cluster overview”, all clusters known to Uyuni/SUSE Manager will be displayed. When a cluster is selected, a new page listing all the details of the cluster will be shown, containing:
+
+- List of all nodes of the cluster - by invoking the cluster provider manager (via Salt) on the management node:
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.list_nodes <cluster name>
+  ```
+
+The cluster provider manager can be implemented as a [virtual module](https://docs.saltstack.com/en/latest/ref/modules/all/index.html) in Salt.
```

Comment:
Do we really need a "virtual module" here or just a "custom module" ?

The "virtual modules" in Salt are like the `pkg` one, which is simply a wrapper of the actual `zypperpkg`, `aptpkg` or whatever actual package manager module depending on the OS.

In the context of this RFC, do we really want a virtual i.a. `clustermgr` module that resolves on the corresponding `shap`, `caasp` or `ses` modules? I guess it's better to not wrap those modules but simply use them as custom modules.

Reply:
@meaksh: custom vs. virtual modules are two different dimensions. The modules will be "custom" as long as they aren't part of upstream Salt (which we would certainly try to fix over time). The idea with the virtual module would simplify the use from SUSE Manager (salt-netapi calls wouldn't have to "know" the name)  and address some of Pau's concerns about having to remember the exact names. But as I've mentioned before, I'm not sure whether we can actually use different virtual modules on the same host (salt minion) as they were designed to be mutually exclusive (e.g. you can have yum _or_ zypper, not both). For the MVP I'd ignore the virtual part for now. We can easily add it later, especially if we also want to use proxy minions.